### PR TITLE
feat(uat): added base source code of future mosquitto-based client

### PR DESCRIPTION
**Pull request by bgklika**
**Issue #, if available:**
Move Mosquitto client sources to auth repo

**Description of changes:**
- Added base structure and some logic of Mosquitto-based MQTT client

**Why is this change necessary:**
We add one more test client

**How was this change tested:**
At the moment manually, client not yet added to CI build.

**Test results**
Control:
```
[INFO ] 2023-05-04 19:19:59.121 [grpc-default-executor-0] GRPCDiscoveryServer - RegisterAgent: agentId agent_mosquitto
[INFO ] 2023-05-04 19:19:59.137 [grpc-default-executor-0] GRPCDiscoveryServer - DiscoveryClient: agentId agent_mosquitto address 127.0.0.1 port 36511
[INFO ] 2023-05-04 19:19:59.140 [grpc-default-executor-0] EngineControlImpl - Created new agent control for agent_mosquitto on 127.0.0.1:36511
[INFO ] 2023-05-04 19:19:59.183 [grpc-default-executor-0] ExampleControl - Agent agent_mosquitto is connected
[INFO ] 2023-05-04 19:19:59.185 [pool-4-thread-1] AgentTestScenario - Playing test scenario for agent id agent_mosquitto
[INFO ] 2023-05-04 19:20:02.819 [pool-4-thread-1] AgentControlImpl - Created connection with id 1 CONNACK 'sessionPresent: false
reasonCode: 0
receiveMaximum: 100
maximumQoS: 1
retainAvailable: true
maximumPacketSize: 149504
wildcardSubscriptionsAvailable: true
subscriptionIdentifiersAvailable: false
sharedSubscriptionsAvailable: true
serverKeepAlive: 60
'
[INFO ] 2023-05-04 19:20:02.820 [pool-4-thread-1] AgentControlImpl - createMqttConnection: MQTT connectionId 1 created
[INFO ] 2023-05-04 19:20:02.820 [pool-4-thread-1] AgentTestScenario - MQTT connection with id 1 is established
[INFO ] 2023-05-04 19:20:07.832 [pool-4-thread-1] AgentControlImpl - SubscribeMqtt: subscribe on connection 1
[INFO ] 2023-05-04 19:20:07.920 [pool-4-thread-1] AgentTestScenario - Subscribe response: connectionId 1 reason codes [0] reason string ''
[INFO ] 2023-05-04 19:20:12.930 [pool-4-thread-1] AgentControlImpl - PublishMqtt: publishing on connectionId 1 topic test/topic
[INFO ] 2023-05-04 19:20:12.942 [pool-4-thread-1] AgentTestScenario - Published connectionId 1 reasonCode 0 reasonString 0
[INFO ] 2023-05-04 19:20:17.949 [pool-4-thread-1] AgentControlImpl - UnsubscribeMqtt: unsubscribe on connectionId 1
[INFO ] 2023-05-04 19:20:17.958 [pool-4-thread-1] AgentTestScenario - Unsubscribe response: connectionId 1 reason codes [] reason string ''
[INFO ] 2023-05-04 19:20:32.979 [pool-4-thread-1] AgentControlImpl - closeMqttConnection: MQTT connectionId 1 closed
[INFO ] 2023-05-04 19:20:32.994 [pool-4-thread-1] AgentControlImpl - shutdown request sent successfully
[INFO ] 2023-05-04 19:20:33.007 [grpc-default-executor-0] GRPCDiscoveryServer - UnregisterAgent: agentId agent_mosquitto reason Agent shutdown by OTF request 'That's it.'
[INFO ] 2023-05-04 19:20:33.012 [grpc-default-executor-0] ExampleControl - Agent agent_mosquitto is disconnected
```
Client:
```
[INFO ] 2023-05-04 19:19:59.121 [grpc-default-executor-0] GRPCDiscoveryServer - RegisterAgent: agentId agent_mosquitto
[INFO ] 2023-05-04 19:19:59.137 [grpc-default-executor-0] GRPCDiscoveryServer - DiscoveryClient: agentId agent_mosquitto address 127.0.0.1 port 36511
[INFO ] 2023-05-04 19:19:59.140 [grpc-default-executor-0] EngineControlImpl - Created new agent control for agent_mosquitto on 127.0.0.1:36511
[INFO ] 2023-05-04 19:19:59.183 [grpc-default-executor-0] ExampleControl - Agent agent_mosquitto is connected
[INFO ] 2023-05-04 19:19:59.185 [pool-4-thread-1] AgentTestScenario - Playing test scenario for agent id agent_mosquitto
[INFO ] 2023-05-04 19:20:02.819 [pool-4-thread-1] AgentControlImpl - Created connection with id 1 CONNACK 'sessionPresent: false
reasonCode: 0
receiveMaximum: 100
maximumQoS: 1
retainAvailable: true
maximumPacketSize: 149504
wildcardSubscriptionsAvailable: true
subscriptionIdentifiersAvailable: false
sharedSubscriptionsAvailable: true
serverKeepAlive: 60
'
[INFO ] 2023-05-04 19:20:02.820 [pool-4-thread-1] AgentControlImpl - createMqttConnection: MQTT connectionId 1 created
[INFO ] 2023-05-04 19:20:02.820 [pool-4-thread-1] AgentTestScenario - MQTT connection with id 1 is established
[INFO ] 2023-05-04 19:20:07.832 [pool-4-thread-1] AgentControlImpl - SubscribeMqtt: subscribe on connection 1
[INFO ] 2023-05-04 19:20:07.920 [pool-4-thread-1] AgentTestScenario - Subscribe response: connectionId 1 reason codes [0] reason string ''
[INFO ] 2023-05-04 19:20:12.930 [pool-4-thread-1] AgentControlImpl - PublishMqtt: publishing on connectionId 1 topic test/topic
[INFO ] 2023-05-04 19:20:12.942 [pool-4-thread-1] AgentTestScenario - Published connectionId 1 reasonCode 0 reasonString 0
[INFO ] 2023-05-04 19:20:17.949 [pool-4-thread-1] AgentControlImpl - UnsubscribeMqtt: unsubscribe on connectionId 1
[INFO ] 2023-05-04 19:20:17.958 [pool-4-thread-1] AgentTestScenario - Unsubscribe response: connectionId 1 reason codes [] reason string ''
[INFO ] 2023-05-04 19:20:32.979 [pool-4-thread-1] AgentControlImpl - closeMqttConnection: MQTT connectionId 1 closed
[INFO ] 2023-05-04 19:20:32.994 [pool-4-thread-1] AgentControlImpl - shutdown request sent successfully
[INFO ] 2023-05-04 19:20:33.007 [grpc-default-executor-0] GRPCDiscoveryServer - UnregisterAgent: agentId agent_mosquitto reason Agent shutdown by OTF request 'That's it.'
[INFO ] 2023-05-04 19:20:33.012 [grpc-default-executor-0] ExampleControl - Agent agent_mosquitto is disconnected
```

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by MikeDombo — COMMENTED** [review]
I thought we were going to use the command line version of mosquitto to avoid needing to compile?

File: uat/custom-components/client-mosquitto-c/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# mosquitto test client
+
+MQTT v5.0 C client for tests based on mosquitto library
+
+## Cloning the repo
+```bash
+
+git clone git@gitlab.klika-tech.com:mzon-ggmq/mosquitto-test-client.git
```
Comment (MikeDombo) [inline uat/custom-components/client-mosquitto-c/README.md:8]: update

**Comment by bgklika** [comment]
> I thought we were going to use the command line version of mosquitto to avoid needing to compile?
mosquitto provides couple of  command line tools to publish and subscribe, but that can't meet requirements defined in gRPC due to completely miss handling of protocol details.

**Comment by MikeDombo** [comment]
Work backwards from the requirement in the test, not the requirement of gRPC.

Does the mosquitto command line implement everything we need from the perspective of the test?

File: uat/custom-components/client-mosquitto-c/src/GRPCLink.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                logd("Local address is %s\n", local_ip.c_str());
+
+                GRPCControlServer * server = new GRPCControlServer(*client, local_ip.c_str(), AUTOSELECT_PORT);
+                if (server) {
+                    unsigned short my_service_port = server->getPort();
+                    if (client->DiscoveryAgent(local_ip.c_str(), my_service_port)) {
+                        m_client = client;
+                        m_server = server;
+                    } else {
+                         throw GRPCException("Couldn't discovery client");  // TODO: add strerror(errno)
```
Comment (MikeDombo) [inline uat/custom-components/client-mosquitto-c/src/GRPCLink.cpp:90]: typo: discover.

File: uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+#include "mqtt_client_control.grpc.pb.h"
+
+#include "MqttConnection.h"                     /* self header */
+#include "MqttException.h"
+#include "logger.h"                             /* logd() */
+
+#define DEFAULT_DISCONNECT_REASON       MQTT_RC_NORMAL_DISCONNECTION
+#define DEFAULT_DISCONNECT_TIMEOUT      10
+
+#define CONNECT_REQUST_ID               (1024*64 + 1)
```
Comment (MikeDombo) [inline uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp:26]: REQUEST

**Comment by bgklika** [comment]
I think not, they are very simplified.

File: uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+void MqttConnection::removePendingRequestLocked(int request_id) {
+
+    auto it = m_requests.find(request_id);
+    if (it != m_requests.end()) {
+        delete it->second;
+        m_requests.erase(it);
+    }
+}
+
+/*
```
Comment (MikeDombo) [inline uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp:476]: remove unused?

File: uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp (hunk 3 above)
Comment (bgklika) [reply uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp]: fixed in 8106e75fbe54a32c4b22de1157aadbf12ffe3a5b

File: uat/custom-components/client-mosquitto-c/src/GRPCLink.cpp (hunk 2 above)
Comment (bgklika) [reply uat/custom-components/client-mosquitto-c/src/GRPCLink.cpp]: fixed in 8106e75fbe54a32c4b22de1157aadbf12ffe3a5b

File: uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp (hunk 4 above)
Comment (bgklika) [reply uat/custom-components/client-mosquitto-c/src/MqttConnection.cpp]: removed in 8106e75fbe54a32c4b22de1157aadbf12ffe3a5b

**Review by MikeDombo — APPROVED** [review]

**Pull request merged**